lib/ecto/adapters/sql.ex

```diff
@@ -245,6 +245,69 @@ defmodule Ecto.Adapters.SQL do
     end
   end
 
+  @doc """
+  Executes an EXPLAIN statement or similar for the given query according to its kind and the
+  adapter in the given repository.
+
+  ## Examples
+
+      # Postgres
+      iex> Ecto.Adapters.SQL.explain(:all, Repo, Post)
+      "Seq Scan on posts p0  (cost=0.00..12.12 rows=1 width=443)"
+
+      # MySQL
+      iex> Ecto.Adapters.SQL.explain(:all, from(p in Post, where: p.title == "title")) |> IO.puts()
+      +----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
+      | id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+      +----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
+      |  1 | SIMPLE      | p0    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |    100.0 | Using where |
+      +----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
+
+  It's safe to execute it for updates and deletes, no data change will be commited:
+
+      iex> Ecto.Adapters.SQL.explain(:update_all, Repo, from(p in Post, update: [set: [title: "new title"]]))
+      "Update on posts p0  (cost=0.00..11.70 rows=170 width=449)\n  ->  Seq Scan on posts p0  (cost=0.00..11.70 rows=170 width=449)"
+
+  Valid `opts` will be mapped directly to the EXPLAIN statement's options for the adapter in use,
+  so please note each adapter may have specific options:
+
+      iex> Ecto.Adapters.SQL.explain(:all, Repo, Post, analyze: true)
+      "Seq Scan on posts p0  (cost=0.00..11.70 rows=170 width=443) (actual time=0.013..0.013 rows=0 loops=1)\nPlanning Time: 0.031 ms\nExecution Time: 0.021 ms"
+
+  This function is also available under the repository with name `explain`:
+
+      iex> Repo.explain(:all, from(p in Post, where: p.title == "title"))
+      "Seq Scan on posts p0  (cost=0.00..12.12 rows=1 width=443)\n  Filter: ((title)::text = 'title'::text)"
+
+  And note that only a textual format is supported at this moment, and you may want to call `IO.puts/1` to properly format the output.
+
+  """
+  @spec explain(pid() | Ecto.Repo.t | Ecto.Adapter.adapter_meta,
+                :all | :update_all | :delete_all,
+                Ecto.Queryable.t, Keyword.t) :: String.t | Exception.t
+  def explain(repo, operation, queryable, opts \\ [])
+
+  def explain(repo, operation, queryable, opts) when is_atom(repo) or is_pid(repo) do
+    explain(Ecto.Adapter.lookup_meta(repo), operation, queryable, opts)
+  end
+
+  def explain(%{repo: repo} = adapter_meta, operation, queryable, opts) do
+    Ecto.Multi.new()
+    |> Ecto.Multi.run(:explain, fn _, _ ->
+      {prepared, params} = to_sql(operation, repo, queryable)
+      sql_call(adapter_meta, :explain_query, [prepared, opts], params, [])
```

Comment:
Turns out I had to separate `params` and `opts` like is done on other functions https://github.com/elixir-ecto/ecto_sql/pull/231/commits/9367b6316298d718b8a7338d4f4d07ac9644537b

Reply:
I am not sure if that's the best way to go. Maybe it would be better to keep it as a single `opts` and then let the adapter take out what it may need in `build_explain_opts` and pass the remaining options to the `query`? I am worried people will get confused and pass the options in the wrong places. :) Since the chance of conflict is low, I would rather keep a single `opts` argument.

Reply:
Sure, makes sense and also it's cleaner.